src/Symfony/Component/Form/Extension/Validator/EventListener/ValidationListener.php

```diff
@@ -35,17 +41,27 @@ public static function getSubscribedEvents()
         return [FormEvents::POST_SUBMIT => 'validateForm'];
     }
 
-    public function __construct(ValidatorInterface $validator, ViolationMapperInterface $violationMapper)
+    public function __construct(ValidatorInterface $validator, ViolationMapperInterface $violationMapper, ?EventDispatcherInterface $eventDispatcher = null)
     {
         $this->validator = $validator;
         $this->violationMapper = $violationMapper;
+
+        if (!$this->eventDispatcher) {
+            @trigger_error(sprintf('The "$eventDispatcher" argument to the "%s" constructor will be required in Symfony 6.0.', self::class));
+        }
+
+        $this->eventDispatcher = $eventDispatcher;
     }
 
     public function validateForm(FormEvent $event)
     {
         $form = $event->getForm();
 
         if ($form->isRoot()) {
+            if ($this->eventDispatcher) {
+                $this->eventDispatcher->dispatch(new PreValidateEvent($event->getForm(), $event->getData()), FormValidationEvents::PRE_VALIDATE);
```

Comment:
is there any reason for these events to be root-only?

Reply:
Validation is only done for root forms, hence the checks here.

Reply:
given i have a custom form type listening to VALIDATE events

now, im nesting this form type somewhere else ...all of a sudden the listener isnt called anymore on the child type.

Im not sure that's expected, given validator service maps violations back to form types, so each type has a PRE/POST validation knowledge imho.

Reply:
Well, the issue is that this is form validation. Form validation is only triggered on root forms. You always have to attach validation listeners to the form that is being validated.

Reply:
form validation bubbles down.

put different, why shouldnt i be able to listen to POST_VALIDATE in a child form to check for some synchronized errors?

Reply:
Form validation does not bubble down. Validation does, in that the form validation metadata contains a `Traverse` constraint, which triggers validation for fields, child forms, and the entity. The POST_SUBMIT event for any non-root forms does not trigger validation, which is why there should be no **form**.(pre|post)_validate event

Reply:
hm yes, i agree. Im  a bit worried the difference is too subtle, compared to current post-submit with prio >0.

Is there any reason to prefer the new events actually?

Reply:
I am kind of interested in this discussion - because my [Bundle](https://github.com/Braunstetter/valid-form-event) does exactly this. It works quite nicely.

The most important parts for me - and my project - were - it has to bubble down - because I want to be able to run code inside of child forms whenever the whole form was successfully validated.

And inside of the $event variable, you have to be able to access the current form (current child) and the root form. Otherwise, it would not be possible to detect what data to use for some action like uploading an image or persisting an entity etc.

Like I said. My bundle works like a charm. I like to see this inside the Symfony core (form component).
But if these success metrics are not given, I will stick to my bundle ;).

Anyways. I like to see that you guys put effort into this.